django/conf/__init__.py

```diff
@@ -140,7 +140,14 @@ def __init__(self, settings_module):
         # store the settings module in case someone later cares
         self.SETTINGS_MODULE = settings_module
 
-        mod = importlib.import_module(self.SETTINGS_MODULE)
+        try:
+            mod = importlib.import_module(self.SETTINGS_MODULE)
+        except ImportError as exc:
+            # If the settings module cannot be imported, treat it as a configuration error.
+            # Inability to import any other module should not be treated as a configuration error.
+            if self.SETTINGS_MODULE == exc.name:
```

Comment:
This will introduce a regression for ticket-28791 when you have a dotted path, e.g.
```
$ export DJANGO_SETTINGS_MODULE=dist.plug.settings
$ django-admin startproject new_test_project
Traceback (most recent call last):
  File "/.virtualenvs/ticket-32915/bin/django-admin", line 33, in <module>
    sys.exit(load_entry_point('Django', 'console_scripts', 'django-admin')())
  File "/django/django/core/management/__init__.py", line 417, in execute_from_command_line
    utility.execute()
  File "/django/django/core/management/__init__.py", line 363, in execute
    settings.INSTALLED_APPS
  File "/django/django/conf/__init__.py", line 69, in __getattr__
    self._setup(name)
  File "/django/django/conf/__init__.py", line 56, in _setup
    self._wrapped = Settings(settings_module)
  File "/django/django/conf/__init__.py", line 144, in __init__
    mod = importlib.import_module(self.SETTINGS_MODULE)
  File "/usr/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 961, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 961, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 973, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'dist'
```
Also it doesn't fix this issue when a settings module is in the same directory, e.g.
```
$ export DJANGO_SETTINGS_MODULE=bad_settings
$ python manage.py runserver --nostatic
usage: manage.py runserver [-h] [--ipv6] [--nothreading] [--noreload] [--version] [-v {0,1,2,3}] [--settings SETTINGS] [--pythonpath PYTHONPATH]
                           [--traceback] [--no-color] [--force-color] [--skip-checks]
                           [addrport]
manage.py runserver: error: unrecognized arguments: --nostatic
```

Reply:
Would it be acceptable to ignore the import error only when running `startproject` and stop execution for all other commands?

Reply:
> Would it be acceptable to ignore the import error only when running `startproject` and stop execution for all other commands?

I don't think so, you can have other custom commands that don't use settings.

Reply:
I'm not sure how to proceed here. Do you have any tips?

Reply:
Not really, sorry :disappointed: it's tricky.

Reply:
Trying to understand the issue some more. My findings so far:

- There are two commands called `runserver`:
    - [`django.core.runserver`](https://github.com/django/django/blob/0250340e372f652c4f276e6874d452d683c94dfe/django/core/management/commands/runserver.py) which does not have `--nostatic` and `--insecure`.
    - [`django.contrib.staticfiles.runserver`](https://github.com/django/django/blob/0250340e372f652c4f276e6874d452d683c94dfe/django/contrib/staticfiles/management/commands/runserver.py) which inherits the core runserver, and adds `--nostatic` and `--insecure`.
- When settings haven't been configured yet, only `django.core` commands are available for execution --> [`get_commands`](https://github.com/django/django/blob/0250340e372f652c4f276e6874d452d683c94dfe/django/core/management/__init__.py#L68-L69).
- In the case of this bug, since settings aren't configured properly, django tries to run `django.core.runserver` and it fails since `--nostatic` is an arg on that command.
- When the settings file is correctly imported it has `django.contrib.staticfiles` in `INSTALLED_APPS` (unless some has removed it) and works with `--nostatic`.
- The following are commands that come from `django.core`:

```json
{"check": "django.core",
 "compilemessages": "django.core",
 "createcachetable": "django.core",
 "dbshell": "django.core",
 "diffsettings": "django.core",
 "dumpdata": "django.core",
 "flush": "django.core",
 "inspectdb": "django.core",
 "loaddata": "django.core",
 "makemessages": "django.core",
 "makemigrations": "django.core",
 "migrate": "django.core",
 "runserver": "django.core",
 "sendtestemail": "django.core",
 "shell": "django.core",
 "showmigrations": "django.core",
 "sqlflush": "django.core",
 "sqlmigrate": "django.core",
 "sqlsequencereset": "django.core",
 "squashmigrations": "django.core",
 "startapp": "django.core",
 "startproject": "django.core",
 "test": "django.core",
 "testserver": "django.core"}
```

- Any commands outside of these (whether they come from Django or are defined in the user's app) cannot be invoked without settings being configured correctly.
- Of the commands listed above, most of them cannot be run without settings being configured correctly.

---

With that in mind, I propose adding a flag like `SETTINGS_OPTIONAL = True` on the commands that can run even in the absence of properly configured settings. Any other command will not be called unless settings is configured correctly. Does this sound good?

Reply:
@felixxm -

- There are two commands called `runserver`:
    - [`django.core.runserver`](https://github.com/django/django/blob/0250340e372f652c4f276e6874d452d683c94dfe/django/core/management/commands/runserver.py) which does not have `--nostatic` and `--insecure`.
    - [`django.contrib.staticfiles.runserver`](https://github.com/django/django/blob/0250340e372f652c4f276e6874d452d683c94dfe/django/contrib/staticfiles/management/commands/runserver.py) which inherits the core runserver, and adds `--nostatic` and `--insecure`.
- When settings haven't been configured yet, only `django.core` commands are available for execution --> [`get_commands`](https://github.com/django/django/blob/0250340e372f652c4f276e6874d452d683c94dfe/django/core/management/__init__.py#L68-L69).
- In the case of this bug, since settings aren't configured properly, django tries to run `django.core.runserver` and it fails since `--nostatic` is an arg on that command.
- When the settings file is correctly imported it has `django.contrib.staticfiles` in `INSTALLED_APPS` (unless some has removed it) and works with `--nostatic`.
- The following are commands that come from `django.core`:

```json
{"check": "django.core",
 "compilemessages": "django.core",
 "createcachetable": "django.core",
 "dbshell": "django.core",
 "diffsettings": "django.core",
 "dumpdata": "django.core",
 "flush": "django.core",
 "inspectdb": "django.core",
 "loaddata": "django.core",
 "makemessages": "django.core",
 "makemigrations": "django.core",
 "migrate": "django.core",
 "runserver": "django.core",
 "sendtestemail": "django.core",
 "shell": "django.core",
 "showmigrations": "django.core",
 "sqlflush": "django.core",
 "sqlmigrate": "django.core",
 "sqlsequencereset": "django.core",
 "squashmigrations": "django.core",
 "startapp": "django.core",
 "startproject": "django.core",
 "test": "django.core",
 "testserver": "django.core"}
```

- Any commands outside of these (whether they come from Django or are defined in the user's app) cannot be invoked without settings being configured correctly.
- Of the commands listed above, most of them cannot be run without settings being configured correctly.

---

With that in mind, I propose:

- Add a flag like `requires_settings_configured = True` on the commands that can run even in the absence of properly configured settings.
- Any other command will not be called unless settings is configured correctly. 
- I will go through all the other `django.core` commands and flag them as well if this approach is approved.
- I have pushed the latest code changes and updated proof in the PR description.

Does this sound good?

Reply:
I believe there should be a way to do it using the original approach by inspecting the traceback, but as @felixxm said it may be tricky.

Reply:
I'm not sure if taking that route is a good idea because it's not necessarily addressing the root cause of the issue. Running `./manage.py runserver` with the wrong DJANGO_SETTINGS_MODULE is calling a whole other command altogether (that happens to have the same name). If the child class had a different name like `static_runserver` then the error message would've been a more obvious `Unknown command: 'static_runserver'` when settings weren't imported rather the somewhat cryptic error that's shown now.

As things stand on master, execution continues despite failing to import settings with the assumption that the command itself will error out if it needs to use settings. The current implementation in this PR only makes the "does not need settings" aspect of it explicit rather than letting execution continue till an error occurs at some other point.

Reply:
I'm not sure I understand your reply or how your approach addresses the issue I had in mind. In the case where `requires_settings_configured` is false, how are you distinguishing between the case of an error due to, say, a missing settings file and the case where a settings file is present but a different kind of error occurs while importing / processing that file?

Reply:
We don't have to distinguish between different kinds of errors at runtime. The only commands that can be run without a proper settings file are those that come from `django.core` ([ref](https://github.com/django/django/pull/14622#discussion_r667949377)). We just need to ensure that django.core commands that don't need settings are flagged correctly. Why the settings file couldn't be imported would not matter at this point.

Reply:
The settings file may not be required, but it is still optional. Consider the case of someone that puts code they want to run with each command as a dependency of their settings file. If any error encountered while importing the settings file is swallowed, then that person's code will silently not run. This is why I was suggesting inspecting the stack trace to distinguish a missing settings file from other types of errors.

Reply:
Got it, makes sense. I will check out how to differentiate between the two cases.

Reply:
@cjerdonek - Hey! I've updated the code and manual testing proof in the PR description. Could you please check it out?

Reply:
@rohithpr Does your proposed solution handle / distinguish the case of importing a nonexistent module from within the settings module? You should also show that, and include in tests when the time comes.

Reply:
@cjerdonek  - Yup, that is handled too.

<img width="810" alt="image" src="https://user-images.githubusercontent.com/10276811/126747368-bafbdb97-55dd-4099-8674-22a71171b9c9.png">